src/vm/eventtrace.cpp

```diff
@@ -1783,18 +1786,60 @@ void BulkTypeEventLogger::FireBulkTypeEvent()
     }
 
     Win32EventWrite(Microsoft_Windows_DotNETRuntimeHandle, &BulkType, iDesc, EventData);
+    
+#else
 
+    UINT iSize = 0;
+    
+    for (int iTypeData = 0; iTypeData < m_nBulkTypeValueCount; iTypeData++)
+    {
+        // Do fixed-size data as one bulk copy
+        memcpy(
+                m_BulkTypeEventBuffer + iSize,
+                &(m_rgBulkTypeValues[iTypeData].fixedSizedData),
+                sizeof(m_rgBulkTypeValues[iTypeData].fixedSizedData));
+        iSize += sizeof(m_rgBulkTypeValues[iTypeData].fixedSizedData);
+
+        // Do var-sized data individually per field
+
+        LPCWSTR wszName = m_rgBulkTypeValues[iTypeData].sName.GetUnicode();
+        if (wszName == NULL)
+        {
+            m_BulkTypeEventBuffer[iSize++] = 0;
+            m_BulkTypeEventBuffer[iSize++] = 0;
+        }
+        else
+        {
+            UINT nameSize = (m_rgBulkTypeValues[iTypeData].sName.GetCount() + 1) * sizeof(WCHAR);
+            memcpy(m_BulkTypeEventBuffer + iSize, wszName, nameSize);
+            iSize += nameSize;
+        }
+
+        // Type parameter count
+        ULONG params = m_rgBulkTypeValues[iTypeData].rgTypeParameters.GetCount();
+        
+        ULONG *ptrInt = (ULONG*)(m_BulkTypeEventBuffer + iSize);
+        *ptrInt = params;
+        iSize += 4;
+        
+        m_rgBulkTypeValues[iTypeData].cTypeParameters = params;
+
+        // Type parameter array
+        if (m_rgBulkTypeValues[iTypeData].cTypeParameters > 0)
+        {
+            memcpy(m_BulkTypeEventBuffer + iSize, m_rgBulkTypeValues[iTypeData].rgTypeParameters.GetElements(), sizeof(ULONGLONG) * m_rgBulkTypeValues[iTypeData].cTypeParameters);
+            iSize += sizeof(ULONGLONG) * m_rgBulkTypeValues[iTypeData].cTypeParameters;
+        }
+    }
```

Comment:
Overall you are using m_rgBulkTypeValues[iTypeData] 10 times. Seems like it is worth taking a reference to the element first thing inside the loop and go with that:
e.g.:
BulkTypeValue& target = m_rgBulkTypeValues[iTypeData];

optimizers are good, if you help them.... and readability improves quite a bit.